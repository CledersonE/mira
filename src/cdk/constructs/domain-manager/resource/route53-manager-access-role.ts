import { Construct } from '@aws-cdk/core'
import {
  Effect,
  ManagedPolicy,
  PolicyStatement
} from '@aws-cdk/aws-iam'
import { MiraConfig } from '../../../../config/mira-config'
import { MiraStack } from '../../../stack'

export const ROUTE53_MANAGER_POLICY_BOUNDARY = 'Route53ManagerPolicyBoundary'

export class Route53ManagerAccessRole extends MiraStack {
  constructor (parent: Construct) {
    super(parent, Route53ManagerAccessRole.name)
    const { hostedZoneId } = MiraConfig.getDomainConfig()
    if (!hostedZoneId) {
      throw new Error('"hostedZoneId" is required in "domain" config')
    }
    new ManagedPolicy(this, 'permissionBoundaryPolicy',
      {
        managedPolicyName: MiraConfig.getBaseStackName('Route53ManagerPolicyBoundary'),
        description: 'Boundary that defines what action can be performed by the Route53Manager stack resources',
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            resources: ['*'],
            actions: [
              'route53:GetChange'
            ]
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            resources: ['*'],
            actions: [
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:PutLogEvents'
            ]
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [`arn:aws:route53:::hostedzone/${hostedZoneId}`],
            actions: [
              'route53:ChangeResourceRecordSets',
              'route53:ListResourceRecordSets'
            ]
          })
        ]
      })
  }
}
