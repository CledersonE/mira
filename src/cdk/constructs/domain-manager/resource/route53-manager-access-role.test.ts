import { SynthUtils } from "@aws-cdk/assert"
import { Stack, App } from "@aws-cdk/core"

import {
  MiraConfig,
  DomainConfig,
  Account,
} from "../../../../config/mira-config"

import { Route53ManagerAccessRole } from "./route53-manager-access-role"

describe("Route53ManagerAccessRole", () => {
  it("throws when hostedZoneId is not set in config", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("Route53Manager"),
      {}
    )

    MiraConfig.getDomainConfig = (): DomainConfig => ({
      accounts: [],
    })

    MiraConfig.getEnvironment = (): Account => ({
      name: "some-name",
      profile: "some-profile",
      env: { account: "12345", region: "eu-west-1" },
    })

    expect(() => new Route53ManagerAccessRole(stack)).toThrowError(
      '"hostedZoneId" is required in "domain" config'
    )
  })

  test("renders output correctly", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("Route53ManagerAccessRole"),
      {}
    )
    MiraConfig.getDomainConfig = (): DomainConfig => ({
      hostedZoneId: 'Z123456789',
      accounts: [
        'account1',
        'domain'
      ],
    })

    MiraConfig.getEnvironment = (envName: string): Account => {
      if (envName === 'domain') {
        return {
          name: "domain",
          profile: "domain",
          env: { account: "domain", region: "eu-west-1" },
        }
      }

      return {
        name: "account1",
        profile: "account1",
        env: { account: "account1", region: "eu-west-1" },
      }
    }
    const rm = new Route53ManagerAccessRole(stack)
    expect(SynthUtils.toCloudFormation(rm)).toMatchSnapshot()
  })
})
