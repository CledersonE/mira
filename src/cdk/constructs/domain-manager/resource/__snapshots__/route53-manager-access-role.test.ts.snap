// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route53ManagerAccessRole renders output correctly 1`] = `
Object {
  "Resources": Object {
    "permissionBoundaryPolicyB4A5FF11": Object {
      "Properties": Object {
        "Description": "Boundary that defines what action can be performed by the Route53Manager stack resources",
        "ManagedPolicyName": "John-MyGreatApp-account1-Route53ManagerPolicyBoundary",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "route53:ChangeResourceRecordSets",
                "route53:ListResourceRecordSets",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:route53:::hostedzone/Z123456789",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
  },
}
`;
